<?xml version = "1.0" ?>
<robot name = "my_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <!-- Constants for robot dimensions -->
  <xacro:property name="PI" value="3.1415926535897931"/>
  <xacro:property name="mass" value="1" /> <!-- arbitrary value for mass -->
  <xacro:property name="width" value="0.1" /> <!-- Square dimensions (widthxwidth) of beams -->
  <xacro:property name="height1" value="1" /> <!-- 1 -->
  <xacro:property name="height2" value="2" /> <!-- 2 -->
  <xacro:property name="height3" value="3" /> <!-- 3 -->
  <xacro:property name="height4" value="4" /> <!-- 4 -->
  <xacro:property name="camera_link" value="0.05" /> <!-- Size of square 'camera' box -->
  <xacro:property name="axel_offset" value="0.05" /> <!-- Space btw top of beam and the each joint -->

  <!-- Import all Gazebo-customization elements, including Gazebo colors -->
  <xacro:include filename="$(find rrbot_description)/urdf/rrbot.gazebo" />
  <!-- Import Rviz colors -->
  <xacro:include filename="$(find rrbot_description)/urdf/materials.xacro" />

  <!-- Used for fixing robot to Gazebo 'base_link' -->
  <link name="world"/>

  <joint name="joint1" type="continuous">
    <parent link="world"/>
    <child link="link1"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>
  
  
  <link name="link1">
    <collision>
      <origin xyz="0 0 ${height1}" rpy="0 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height2}"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 ${height1}" rpy="0 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height2}"/>
      </geometry>
      <material name="orange"/>
    </visual>

    <inertial>
      <origin xyz="0 0 ${height1}" rpy="0 0 0"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 12.0 * (width*width + height1*height1)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height1*height1 + width*width)}" iyz="0.0"
	  izz="${mass / 12.0 * (width*width + width*width)}"/>
    </inertial>
  </link>
  
  <joint name="joint_fixed" type="fixed">
    <parent link="link1"/>
    <child link="link2"/>
    <origin xyz="0 0 ${height2}" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
  </joint>
  
  <link name="link2">
    <collision>
      <origin xyz="0 ${height1} ${height2} " rpy="${-PI/2} 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height2}"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 ${height1} ${height2}" rpy="${-PI/2} 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height2}"/>
      </geometry>
      <material name="orange"/>
    </visual>

    <inertial>
      <origin xyz="0 ${height1} ${height2}" rpy="${-PI/2} 0 0"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 12.0 * (width*width + height1*height1)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height1*height1 + width*width)}" iyz="0.0"
	  izz="${mass / 12.0 * (width*width + width*width)}"/>
    </inertial>
  </link>
  
  <joint name="joint2" type="continuous">
    <parent link="link2"/>
    <child link="link3"/>
    <origin xyz="0 ${height2} ${height2}" rpy="0 0 0"/>
    <axis xyz="0 0 1"/>
  </joint>
  
  <link name="link3">
    <collision>
      <origin xyz="0 ${height3} ${height2}" rpy="${-PI/2} 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height2}"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 ${height3} ${height2}" rpy="${-PI/2} 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height2}"/>
      </geometry>
      <material name="orange"/>
    </visual>

    <inertial>
      <origin xyz="0 ${height3} ${height2}" rpy="${-PI/2} 0 0"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 12.0 * (width*width + height1*height1)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 12.0 * (height1*height1 + width*width)}" iyz="0.0"
	  izz="${mass / 12.0 * (width*width + width*width)}"/>
    </inertial>
  </link>
  
  <joint name="joint3" type="prismatic">
    <parent link="link3"/>
    <child link="link4"/>
    <limit effort="100.0" lower="100.0" upper="100.0" velocity="0.5"/>
    <origin xyz="0 ${height4} ${height2}" rpy="0 0 0"/>
    <axis xyz="0 0 -1"/>
  </joint>
  
  <link name="link4">
    <collision>
      <origin xyz="0 ${height4} ${height1}" rpy="${PI} 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height1}"/>
      </geometry>
    </collision>
    
    <visual>
      <origin xyz="0 ${height4} ${height1}" rpy="${PI} 0 0"/>
      <geometry>
	<box size="${width} ${width} ${height1}"/>
      </geometry>
      <material name="orange"/>
    </visual>

    <inertial>
      <origin xyz="0 ${height4} ${height1}" rpy="${PI} 0 0"/>
      <mass value="${mass}"/>
      <inertia
	  ixx="${mass / 24.0 * (width*width + height2*height2)}" ixy="0.0" ixz="0.0"
	  iyy="${mass / 24.0 * (height2*height2 + width*width)}" iyz="0.0"
	  izz="${mass / 24.0 * (width*width + width*width)}"/>
    </inertial>
  </link>
  
  <!-- Turning self - collision on for the entire robot -->
  <gazebo>
        <self_collide>1</self_collide>
  </gazebo>
  
  <!-- ROS Control Plugin for Gazebo -->
  <gazebo>
    <plugin name="gazebo_ros_control" filename ="libgazebo_ros_control.so">
      <robotNamespace>/simple_model</robotNamespace>
    </plugin>
  </gazebo>
  
  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor1">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>
    
  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor2">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <actuator name="motor3">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
    <joint name="joint3">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
  </transmission>
  
</robot>
